### PR TITLE
feat: arrow modifier

### PR DESCRIPTION
**Pull request by atomiks**
I hope this approach is good, need some help with the types

### CODE REVIEW

File: src/types.js [hunk 1]
```diff
@@ -30,6 +30,7 @@ export type State = {|
   elements: {|
     reference: HTMLElement,
     popper: HTMLElement,
+    arrow: HTMLElement,
```
Comment (FezVrasta) [inline src/types.js]: `arrow?: HTMLElement`

File: src/modifiers/arrow.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  // Make sure the arrow doesn't overflow the popper if the center point is
+  // outside of the popper bounds
+  center = within(
+    0,
+    center,
+    state.measures.popper[len] - arrowElementRect[len]
+  );
+
+  // Flow: How to use computed property like {[axis]: center}?
+  state.modifiersData.arrowOffsets = { x: 0, y: 0 };
```
Comment (FezVrasta) [inline src/modifiers/arrow.js]: The property should have the same name of the modifier

File: src/types.js (hunk 1 above)
Comment (atomiks) [reply src/types.js]: We'll need to account for that in `applyStyles` I guess (I did try that)

File: src/modifiers/arrow.js (hunk 2 above)
Comment (atomiks) [reply src/modifiers/arrow.js]: Hmm so should we separate it then like `popperOffsets`? As `computeStyles` assumes `arrowOffsets` name. IIRC some people didn't like how variation placements centered the arrow relative to the reference (they wanted it to be at the start/end always) - they could adjust it there?

File: src/modifiers/applyStyles.js [hunk 3]
```diff
@@ -13,6 +13,11 @@ export function applyStyles(state: State) {
     const attributes = data.attributes[name] || {};
     const element = state.elements[name];
 
+    // arrow is optional
+    if (!element) {
```
Comment (FezVrasta) [inline src/modifiers/applyStyles.js]: Why is this needed? `state.elements.arrow` should not exist if the arrow modifier didn't run.

File: src/modifiers/arrow.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  // outside of the popper bounds
+  center = within(
+    0,
+    center,
+    state.measures.popper[len] - arrowElementRect[len]
+  );
+
+  // Flow: How to use computed property like {[axis]: center}?
+  state.modifiersData.arrow = {};
+  state.modifiersData.arrow[axis] = center;
```
Comment (FezVrasta) [inline src/modifiers/arrow.js]:
```js
  state.modifiersData.arrow = {
    x: 0,
    y: 0,
    [(axis: string)]: center,
  };
```

File: src/modifiers/applyStyles.js (hunk 3 above)
Comment (atomiks) [reply src/modifiers/applyStyles.js]: It's for the types to stop complaining

Comment (FezVrasta) [reply src/modifiers/applyStyles.js]: 🤦‍♂️

File: src/modifiers/arrow.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+  // Make sure the arrow doesn't overflow the popper if the center point is
+  // outside of the popper bounds
+  center = within(
+    0,
+    center,
+    state.measures.popper[len] - arrowElementRect[len]
+  );
+
+  // Prevents breaking syntax highlighting...
```
Comment (FezVrasta) [inline src/modifiers/arrow.js]: Are you using [Babel JavaScript](https://marketplace.visualstudio.com/items?itemName=mgmcdermott.vscode-language-babel) syntax? That supports this kind of annotations.

Comment (atomiks) [reply src/modifiers/arrow.js]: I don't really want to have to change syntax highlighter packages, I already have to turn off a setting while developing with Flow in VSCode :-(

**Comment by atomiks** [comment]
A good thing about this approach is it won't ever affect the Popper's position. In v1, the arrow can affect the Popper's position which causes weird bugs :-(

**Comment by FezVrasta** [comment]
Shouldn't the arrow attempt to stay centered within the popper element rather than the reference?

image

**Comment by atomiks** [comment]
No the whole point is that it stays centered relative to the reference element, that's how it works in v1. 😅 

example

**Comment by FezVrasta** [comment]
But if the reference is overflowing, the other way around doesn't look nicer? 🌈

**Comment by atomiks** [comment]
Yeah but how would that work? It's centered to the reference until it starts overflowing, then "jerks" down suddenly to center in the popper? 

Seems complex to handle regardless. :-\ I think the current v1 behavior has been good enough anyway

**Comment by atomiks** [comment]
Thought about that a bit more. I'm guessing it would only work if the popper is smaller than the reference along that axis - it would look better to keep the arrow centered when the reference is overflowing 🤔

**Comment by atomiks** [comment]
We need to know the `tether` value from `preventOverflow` modifier. How do we access options from other modifiers?

As we know, `tether` allows the popper to overflow if it's about to detach from the reference element.

If we envision the following values:

```js
{
  // the popper never overflows, always stays visible
  tether: false,
  // default is "center" - this should work well for most cases
  tether: true,
  // popper starts overflowing once it reaches the edge of the ref
  // this works well if the ref is like a line of text or something
  tether: "edge",
  // popper starts overflowing once the surfaces of the boxes touch
  // this is how it currently works
  tether: "surface"
}
```

The `arrow` modifier will need to know these to decide what to do.

**Comment by FezVrasta** [comment]
I added a functional test for it, and merged. I will improve the behavior later.
